synapse/handlers/e2e_keys.py

```diff
@@ -984,56 +980,47 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
+        # store it just in case.
+        if (
+            key is None
+            and not self.is_mine(user)
+            # We only get "master" and "self_signing" keys from remote servers
+            and desired_key_type in ["master", "self_signing"]
+        ):
             try:
-                remote_result = yield self.federation.query_client_keys(
-                    user.domain, {"device_keys": {user_id: []}}, timeout=10 * 1000
+                remote_result = yield self.federation.query_user_devices(
+                    user.domain, user_id
                 )
 
-                # Process the result
-                for remote_key_type, remote_user_dict in remote_result.items():
-                    # The key_type variable passed to this function is in the form
-                    # "self_signing","master" etc. whereas the results returned from
-                    # the remote server use "self_signing_keys", "master_keys" etc.
-                    # Remove the "_keys" from the key type
-                    if remote_key_type.endswith("_keys"):
-                        remote_key_type = remote_key_type[:-5]
-
-                    # remote_user_dict is a dictionary in the form of
-                    # {
-                    #   "user_id": {
-                    #     "master_keys": ...
-                    #   },
-                    #   ...
-                    # }
-
-                    # Only extract the keys that pertain to the requested user
-                    key_content_list = remote_user_dict.get(user_id, {}).values()
-
-                    for key_content in key_content_list:
-                        # If the key_type here matches the key we're requesting,
-                        # then this is the key we want to return
-                        if remote_key_type == key_type:
-                            key = key_content
-
-                        # At the same time, save the key to the database for subsequent
-                        # queries
-                        yield self.store.set_e2e_cross_signing_key(
-                            user_id, remote_key_type, key_content
-                        )
-            except (
-                HttpResponseException,
-                NotRetryingDestination,
-                FederationDeniedError,
-                RequestSendFailed,
-            ) as e:
+                # Process each of the retrieved cross-signing keys
+                for key_type in ["master", "self_signing"]:
+                    key_content = remote_result.get(key_type + "_key")
+                    if not key_content:
+                        continue
+
+                    # If this is the desired key type, return it
+                    if key_type == desired_key_type:
+                        key = key_content
+
+                    # At the same time, store this key in the db for
+                    # subsequent queries
+                    yield self.store.set_e2e_cross_signing_key(
```

Comment:
this doesn't look like a thing we can run on a worker (and I think this method is used on workers?)

Reply:
Hmm, I'd have thought so but [workers.md](https://github.com/matrix-org/synapse/blob/master/docs/workers.md) doesn't seem to cover `/_matrix/client/r0/keys/signatures/upload`, so perhaps it is running on master at the moment?

The method which is failing to find the DM room between the two users is running on the federation_reader though.

Reply:
oh, right, yes, we're uploading a signature here. That does sound like a thing that needs to happen on the master. Fair enough. (Though for me it raises even more questions about why the master doesn't know that the shared room exists.)

Reply:
I think you figured it out already, but that race is happening here:

https://github.com/matrix-org/synapse/blob/40042dec0d4bab92ea82122a738c8c495a3adcdd/synapse/handlers/e2e_keys.py#L1211-L1215

when we get signing key updates from the remote server. That code happens on federation_reader.